File: services/marketplace_fulfillment.py
"""What a buyer must tell PulseSoc before paying, decided per order type.

A marketplace listing is one of five types — physical, digital, service, event,
booking — and each needs a genuinely different answer before a card is charged.
A hoodie needs somewhere to ship to. A PDF needs nothing at all. A consultation
needs a date, a time and a timezone. Asking every buyer the same question is
wrong in both directions: it demands a postal address from someone downloading a
file, and it collects nothing from someone booking a call, leaving the seller to
chase them afterwards.

Before this module the checkout screen recognised four values —
``digital``/``pickup``/``shipping``/``both`` — and anything it did not recognise
fell through to ``shipping``. A ``booking`` listing therefore reached the payment
sheet with address collection switched on, so the one order type that never ships
anywhere was the one asking for a delivery address.

Two things live here, and they are deliberately together:

* :func:`resolve_kind` — the single rule that turns a stored listing row into a
  canonical fulfilment kind. The routes call it against the *database* row, never
  against anything the client sent, so the order type is not something a request
  can assert.
* :func:`required_fields` / :func:`validate_details` — what that kind obliges the
  buyer to supply, and whether a submission satisfies it.

Field *keys* are owned here because they are a validation contract. Field
*labels* are not: those are the client's, so they can be translated.

``mobile-native/src/api/marketplaceFulfillment.ts`` is a port of the same two
rules, so the form the buyer fills in is the form the server will accept.
``tests/test_marketplace_fulfillment.py`` pins the two together.
"""

from __future__ import annotations

import os
import re
from typing import Any

# Canonical fulfilment kinds. `choice` variants mean the seller offers more than
# one and the buyer has not answered yet; they never survive into a charge.
KINDS = (
    "shipping",
    "pickup",
    "shipping_or_pickup",
    "digital",
    "service_remote",
    "service_in_person",
    "service_choice",
    "event_online",
    "event_in_person",
    "booking_remote",
    "booking_in_person",
)

#: Kinds that still need the buyer to pick a lane before anything else is asked.
UNDECIDED_KINDS = frozenset({"shipping_or_pickup", "service_choice"})

#: Kinds that hold no physical stock, so no inventory unit is reserved.
STOCKLESS_KINDS = frozenset({
    "digital", "service_remote", "service_in_person", "service_choice",
    "event_online", "event_in_person", "booking_remote", "booking_in_person",
})

DETAILS_REQUIRED_CODE = "FULFILLMENT_DETAILS_REQUIRED"
LANE_REQUIRED_CODE = "FULFILLMENT_REQUIRED"

_PICKUP_OR_SHIPPING = {"both", "pickup_or_shipping", "shipping_or_pickup"}

_ISO_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
_HHMM_RE = re.compile(r"^([01]\d|2[0-3]):[0-5]\d$")
_TZ_RE = re.compile(r"^[A-Za-z0-9_+\-/]{1,64}$")
_COUNTRY_RE = re.compile(r"^[A-Za-z]{2}$")
_TAG_RE = re.compile(r"<[^>]+>")
_WS_RE = re.compile(r"\s+")

# Addresses in these countries are complete without a postal code. Demanding one
# universally is the classic way a checkout built in the US rejects a valid
# address in Ireland or Hong Kong.
_NO_POSTAL_CODE = frozenset({
    "AE", "AO", "AG", "AW", "BS", "BZ", "BJ", "BW", "BF", "BI", "CM", "CF",
    "KM", "CG", "CD", "CK", "CI", "DJ", "DM", "GQ", "ER", "FJ", "TF", "GM",
    "GH", "GD", "GY", "HK", "IE", "JM", "KE", "KI", "KP", "LY", "MO", "MW",
    "ML", "MR", "MU", "MS", "NR", "AN", "NU", "QA", "RW", "KN", "LC", "ST",
    "SC", "SL", "SB", "SO", "SR", "SY", "TZ", "TL", "TK", "TO", "TT", "TV",
    "UG", "VU", "YE", "ZW",
})

# Countries where a carrier will not deliver on street and city alone. Kept
# short on purpose: a region demanded where none exists is a dead end for the
# buyer, and every entry beyond these is an optional line the seller can read.
_REGION_REQUIRED = frozenset({"US", "CA", "AU", "IN", "BR", "MX", "MY", "CN", "AR", "ID"})


def _clean(value: Any, limit: int) -> str:
    text = _TAG_RE.sub(" ", str(value if value is not None else ""))
    return _WS_RE.sub(" ", text).strip()[: int(limit)]


def shipping_countries() -> tuple[str, ...]:
    """Countries the marketplace will accept a delivery address in.

    Shares ``MARKETPLACE_SHIPPING_COUNTRIES`` with the Stripe address-collection
    parameter, so PulseSoc cannot accept an address at review that Stripe would
    have refused at payment.
    """
    raw = os.getenv("MARKETPLACE_SHIPPING_COUNTRIES", "US")
    codes = tuple(c.strip().upper() for c in raw.split(",") if c.strip())
    return codes or ("US",)


# ---------------------------------------------------------------------------
# Kind resolution
# ---------------------------------------------------------------------------

def resolve_kind(listing_type: Any, delivery_type: Any, metadata: Any = None) -> str:
    """Canonical fulfilment kind for a stored listing row.

    Reads the seller's own declarations — the listing type and the type-specific
    metadata they filled in — rather than the delivery column alone, which for a
    service or booking row carries no delivery meaning at all.
    """
    meta = metadata if isinstance(metadata, dict) else {}
    kind = str(listing_type or "").strip().lower()
    delivery = str(delivery_type or "").strip().lower()

    if kind == "digital":
        return "digital"

    if kind == "service":
        location = str(meta.get("service_location") or "").strip().lower()
        if location == "both":
            return "service_choice"
        return "service_in_person" if location == "in_person" else "service_remote"

    if kind == "event":
        venue = str(meta.get("venue_mode") or "").strip().lower()
        return "event_in_person" if venue == "in_person" else "event_online"

    if kind == "booking":
        return "booking_in_person" if str(meta.get("meeting_mode") or "").strip().lower() == "in_person" else "booking_remote"

    # Physical, and anything legacy that never declared a type.
    option = delivery or str(meta.get("delivery_options") or "").strip().lower()
    if option == "digital":
        return "digital"
    if option in _PICKUP_OR_SHIPPING:
        return "shipping_or_pickup"
    if option == "pickup":
        return "pickup"
    return "shipping"


def resolve_choice(kind: str, chosen: Any) -> tuple[str, str]:
    """Settle an undecided kind with the buyer's answer.

    Returns ``(kind, error_code)``. The error is non-empty when the buyer has to
    answer and has not: defaulting on their behalf is how someone collecting in
    person ends up typing an address they never needed.
    """
    if kind not in UNDECIDED_KINDS:
        return kind, ""
    answer = str(chosen or "").strip().lower()
    allowed = {
        "shipping_or_pickup": {"shipping", "pickup"},
        "service_choice": {"service_remote", "service_in_person", "remote", "in_person"},
    }[kind]
    if answer not in allowed:
        return kind, LANE_REQUIRED_CODE
    if kind == "service_choice":
        return ("service_in_person" if answer in {"in_person", "service_in_person"} else "service_remote"), ""
    return answer, ""


def needs_shipping_address(kind: str) -> bool:
    """Whether this kind puts a parcel or a person at a street address."""
    return kind in {"shipping", "service_in_person", "booking_in_person"}


# ---------------------------------------------------------------------------
# Buyer field requirements
# ---------------------------------------------------------------------------

_CONTACT = (("contact_name", "name", True), ("contact_phone", "phone", False))
_CONTACT_WITH_PHONE = (("contact_name", "name", True), ("contact_phone", "phone", True))
_ADDRESS = (
    ("address_line1", "text", True),
    ("address_line2", "text", False),
    ("address_city", "text", True),
    ("address_region", "text", False),
    ("address_postal_code", "text", False),
    ("address_country", "country", True),
)
_WHEN = (("scheduled_date", "date", True), ("scheduled_time", "time", True), ("timezone", "timezone", True))
_NOTES = (("notes", "multiline", False),)

#: kind -> ordered ``(key, type, required)`` triples. Absence is meaningful:
#: ``digital`` asks for nothing because the buyer's account already carries
#: everywhere the file has to go.
_FIELDS: dict[str, tuple[tuple[str, str, bool], ...]] = {
    "digital": (),
    "shipping": (*_CONTACT, *_ADDRESS, ("delivery_notes", "multiline", False)),
    "pickup": (*_CONTACT_WITH_PHONE, ("pickup_preference", "text", False)),
    "service_remote": (*_CONTACT, *_WHEN, *_NOTES),
    "service_in_person": (*_CONTACT_WITH_PHONE, *_WHEN, *_ADDRESS, *_NOTES),
    "event_online": (("attendee_name", "name", True), ("ticket_type", "choice", True)),
    "event_in_person": (("attendee_name", "name", True), ("contact_phone", "phone", False), ("ticket_type", "choice", True)),
    "booking_remote": (*_CONTACT, *_WHEN, *_NOTES),
    "booking_in_person": (*_CONTACT_WITH_PHONE, *_WHEN, *_ADDRESS, *_NOTES),
}

_LIMITS = {"name": 80, "phone": 32, "text": 120, "multiline": 400, "country": 2, "choice": 60, "timezone": 64}


def ticket_options(metadata: Any) -> tuple[str, ...]:
    """Ticket names the seller published, for the one ``choice`` field there is."""
    meta = metadata if isinstance(metadata, dict) else {}
    tickets = meta.get("tickets")
    if not isinstance(tickets, list):
        return ()
    names = tuple(_clean(t.get("name"), 60) for t in tickets if isinstance(t, dict) and _clean(t.get("name"), 60))
    return names


def required_fields(kind: str, metadata: Any = None) -> tuple[str, ...]:
    """Keys the buyer must always fill for ``kind``, in the order they are asked.

    Region and postal code are absent even for a shipping order: whether they are
    required depends on the country, which is not known until the buyer picks
    one. :func:`validate_details` settles those two afterwards.
    """
    out = []
    for key, _type, required in _FIELDS.get(kind, ()):
        if not required:
            continue
        # A seller who published no ticket tiers is selling one undifferentiated
        # admission, so there is nothing for the buyer to choose between.
        if key == "ticket_type" and not ticket_options(metadata):
            continue
        out.append(key)
    return tuple(out)


def field_spec(kind: str, metadata: Any = None) -> tuple[dict[str, Any], ...]:
    """Every field for ``kind``, required or not, with its type and options."""
    tickets = ticket_options(metadata)
    spec = []
    for key, field_type, required in _FIELDS.get(kind, ()):
        if key == "ticket_type":
            if not tickets:
                continue
            spec.append({"key": key, "type": "choice", "required": True, "options": list(tickets)})
            continue
        spec.append({"key": key, "type": field_type, "required": required})
    return tuple(spec)


# ---------------------------------------------------------------------------
# Validation
# ---------------------------------------------------------------------------

def _reject(message: str, field: str = "") -> dict[str, Any]:
    return {"code": DETAILS_REQUIRED_CODE, "status": 400, "message": message, "field": field}


def validate_details(kind: str, raw: Any, metadata: Any = None) -> tuple[bool, Any]:
    """Check a buyer submission against what ``kind`` requires.

    Returns ``(True, cleaned_dict)`` or ``(False, error_dict)``, the error shaped
    like the other pre-flight refusals — ``code``/``status``/``message`` — so the
    three checkout lanes answer it the way they answer everything else.

    Unknown keys are dropped rather than rejected: a client that asks for one
    optional field too many should not be able to fail a buyer's checkout.
    """
    if kind in UNDECIDED_KINDS:
        return False, {"code": LANE_REQUIRED_CODE, "status": 400,
                       "message": "Choose how you want this order fulfilled before you pay.", "field": ""}
    spec = field_spec(kind, metadata)
    if not spec:
        return True, {}
    submitted = raw if isinstance(raw, dict) else {}
    cleaned: dict[str, Any] = {}

    for field in spec:
        key = field["key"]
        value = _clean(submitted.get(key), _LIMITS.get(field["type"], 120))
        if not value:
            if field["required"]:
                return False, _reject(f"{_LABELS[key]} is required before you can pay.", key)
            continue

        field_type = field["type"]
        if field_type == "date":
            if not _ISO_DATE_RE.match(value):
                return False, _reject(f"{_LABELS[key]} must be a calendar date.", key)
        elif field_type == "time":
            if not _HHMM_RE.match(value):
                return False, _reject(f"{_LABELS[key]} must be a time of day.", key)
        elif field_type == "timezone":
            if not _TZ_RE.match(value):
                return False, _reject("Choose a valid timezone.", key)
        elif field_type == "country":
            value = value.upper()
            if not _COUNTRY_RE.match(value):
                return False, _reject("Choose a country.", key)
            if value not in shipping_countries():
                return False, _reject("This seller does not deliver to that country yet.", key)
        elif field_type == "choice":
            if value not in set(field.get("options") or ()):
                return False, _reject(f"Choose {_LABELS[key].lower()}.", key)
        cleaned[key] = value

    if needs_shipping_address(kind):
        country = cleaned.get("address_country", "")
        # Required-ness that depends on the country, resolved after it is known.
        if country in _REGION_REQUIRED and not cleaned.get("address_region"):
            return False, _reject("State or province is required for that country.", "address_region")
        if country not in _NO_POSTAL_CODE and not cleaned.get("address_postal_code"):
            return False, _reject("Postal code is required for that country.", "address_postal_code")

    return True, cleaned


_LABELS = {
    "contact_name": "Full name",
    "contact_phone": "Phone number",
    "attendee_name": "Attendee name",
    "address_line1": "Street address",
    "address_line2": "Apartment or unit",
    "address_city": "City",
    "address_region": "State or province",
    "address_postal_code": "Postal code",
    "address_country": "Country",
    "scheduled_date": "Date",
    "scheduled_time": "Time",
    "timezone": "Timezone",
    "ticket_type": "Ticket type",
    "notes": "Notes",
    "delivery_notes": "Delivery notes",
    "pickup_preference": "Pickup preference",
}


# ---------------------------------------------------------------------------
# Persistence
# ---------------------------------------------------------------------------

def snapshot(kind: str, cleaned: Any) -> dict[str, Any]:
    """The fulfilment record to freeze onto the transaction.

    Written into ``seller_transactions.metadata_json`` beside the commercial
    quote, so an order read back next year still says where it was going and
    when it was booked for, even if the seller has since edited the listing.
    """
    return {"kind": kind, "details": dict(cleaned or {})}


def stripe_shipping(cleaned: Any) -> dict[str, Any]:
    """Stripe's ``shipping`` object for an address PulseSoc already collected.

    Handing it over is what stops the payment sheet asking a second time for
    something the buyer typed on the review screen a moment earlier.
    """
    details = cleaned if isinstance(cleaned, dict) else {}
    if not details.get("address_line1"):
        return {}
    address = {
        "line1": details.get("address_line1", ""),
        "city": details.get("address_city", ""),
        "country": details.get("address_country", ""),
    }
    for key, target in (("address_line2", "line2"), ("address_region", "state"), ("address_postal_code", "postal_code")):
        if details.get(key):
            address[target] = details[key]
    shipping = {"name": details.get("contact_name") or details.get("attendee_name") or "", "address": address}
    if details.get("contact_phone"):
        shipping["phone"] = details["contact_phone"]
    return shipping if shipping["name"] else {}

File: mobile-native/src/api/marketplaceFulfillment.ts
/**
 * What the buyer is asked before paying, decided per order type.
 *
 * A port of `services/marketplace_fulfillment.py` — the same kind resolution and
 * the same field list — so the form rendered here is the form the server will
 * accept. The server re-derives both from the stored listing row and validates
 * again; this copy exists so the buyer sees the right questions immediately
 * rather than discovering them as a rejection.
 *
 * `tests/test_marketplace_fulfillment.py` and this module's jest test pin the
 * two together: a kind added on one side without the other fails there.
 */
import type { MarketplaceListing } from "./marketplace";

export type MarketplaceFulfillmentKind =
  | "shipping"
  | "pickup"
  | "shipping_or_pickup"
  | "digital"
  | "service_remote"
  | "service_in_person"
  | "service_choice"
  | "event_online"
  | "event_in_person"
  | "booking_remote"
  | "booking_in_person";

export const UNDECIDED_KINDS: MarketplaceFulfillmentKind[] = ["shipping_or_pickup", "service_choice"];

export type FulfillmentFieldType =
  | "name"
  | "phone"
  | "text"
  | "multiline"
  | "country"
  | "date"
  | "time"
  | "timezone"
  | "choice";

export type FulfillmentField = {
  key: string;
  type: FulfillmentFieldType;
  required: boolean;
  label: string;
  options?: string[];
};

/**
 * Every spelling of a delivery lane that has ever been stored, folded onto the
 * four this module reasons about. Mirrors `_LANE_WORDS` on the server.
 */
const LANE_WORDS: Record<string, "pickup" | "shipping" | "both" | "digital"> = {
  pickup: "pickup",
  local: "pickup",
  meetup: "pickup",
  shipping: "shipping",
  delivery: "shipping",
  both: "both",
  pickup_or_shipping: "both",
  shipping_or_pickup: "both",
  digital: "digital",
  download: "digital"
};

/**
 * The delivery lane the seller declared, or `""` if they declared none.
 *
 * `delivery_type` does not hold one. Every backend writer of that column stores
 * the *product type* in it — the publish route's INSERT lists
 * `delivery_type, product_type` against `product_type, product_type`, the CJ
 * importer hardcodes `'physical','physical'`, and the column's own default is
 * `'digital'` — so it reads `physical` for every physical listing in the table
 * and it can never read `pickup`, `shipping` or `both`.
 *
 * The seller's actual answer is `listing_metadata.delivery_options`, validated
 * server-side against exactly those three words, and it is read first here. The
 * column is consulted only for a row that declared no listing type at all,
 * where it is the only signal there is.
 *
 * Reading the column first is what made this module answer `shipping` for every
 * physical listing — one tap after the product screen printed "Local pickup"
 * from the metadata, and while the same screen's grid card offered no buy
 * button at all because `physical` matched none of its lane words either.
 */
export function deliveryLane(listing: MarketplaceListing): "" | "pickup" | "shipping" | "both" | "digital" {
  const metadata = (listing.listing_metadata || {}) as Record<string, unknown>;
  const declared = LANE_WORDS[String(metadata.delivery_options || "").trim().toLowerCase()];
  if (declared) return declared;
  if (String(listing.listing_type || listing.product_type || "").trim()) return "";
  return LANE_WORDS[String(listing.delivery_type || "").trim().toLowerCase()] || "";
}

export function resolveFulfillmentKind(listing: MarketplaceListing): MarketplaceFulfillmentKind {
  const metadata = (listing.listing_metadata || {}) as Record<string, unknown>;
  const kind = String(listing.listing_type || listing.product_type || "").trim().toLowerCase();

  if (kind === "digital") return "digital";
  if (kind === "service") {
    const location = String(metadata.service_location || "").trim().toLowerCase();
    if (location === "both") return "service_choice";
    return location === "in_person" ? "service_in_person" : "service_remote";
  }
  if (kind === "event") {
    return String(metadata.venue_mode || "").trim().toLowerCase() === "in_person"
      ? "event_in_person"
      : "event_online";
  }
  if (kind === "booking") {
    return String(metadata.meeting_mode || "").trim().toLowerCase() === "in_person"
      ? "booking_in_person"
      : "booking_remote";
  }

  // Physical, and anything legacy that never declared a type.
  const lane = deliveryLane(listing);
  if (lane === "digital") return "digital";
  if (lane === "both") return "shipping_or_pickup";
  if (lane === "pickup") return "pickup";
  return "shipping";
}

/** Settle an undecided kind with the lane the buyer picked, or return it unchanged. */
export function resolveFulfillmentChoice(
  kind: MarketplaceFulfillmentKind,
  chosen: string
): MarketplaceFulfillmentKind | null {
  if (!UNDECIDED_KINDS.includes(kind)) return kind;
  const answer = String(chosen || "").trim().toLowerCase();
  if (kind === "shipping_or_pickup") {
    return answer === "shipping" || answer === "pickup" ? (answer as MarketplaceFulfillmentKind) : null;
  }
  if (answer === "in_person" || answer === "service_in_person") return "service_in_person";
  if (answer === "remote" || answer === "service_remote") return "service_remote";
  return null;
}

export function fulfillmentNeedsAddress(kind: MarketplaceFulfillmentKind) {
  return kind === "shipping" || kind === "service_in_person" || kind === "booking_in_person";
}

const LABELS: Record<string, string> = {
  contact_name: "Full name",
  contact_phone: "Phone number",
  attendee_name: "Attendee name",
  address_line1: "Street address",
  address_line2: "Apartment or unit",
  address_city: "City",
  address_region: "State or province",
  address_postal_code: "Postal code",
  address_country: "Country",
  scheduled_date: "Date",
  scheduled_time: "Time",
  timezone: "Timezone",
  ticket_type: "Ticket type",
  notes: "Notes",
  delivery_notes: "Delivery notes",
  pickup_preference: "Pickup preference",
};

type Triple = [string, FulfillmentFieldType, boolean];

const CONTACT: Triple[] = [
  ["contact_name", "name", true],
  ["contact_phone", "phone", false],
];
const CONTACT_WITH_PHONE: Triple[] = [
  ["contact_name", "name", true],
  ["contact_phone", "phone", true],
];
const ADDRESS: Triple[] = [
  ["address_line1", "text", true],
  ["address_line2", "text", false],
  ["address_city", "text", true],
  ["address_region", "text", false],
  ["address_postal_code", "text", false],
  ["address_country", "country", true],
];
const WHEN: Triple[] = [
  ["scheduled_date", "date", true],
  ["scheduled_time", "time", true],
  ["timezone", "timezone", true],
];
const NOTES: Triple[] = [["notes", "multiline", false]];

const FIELDS: Record<MarketplaceFulfillmentKind, Triple[]> = {
  digital: [],
  shipping: [...CONTACT, ...ADDRESS, ["delivery_notes", "multiline", false]],
  pickup: [...CONTACT_WITH_PHONE, ["pickup_preference", "text", false]],
  shipping_or_pickup: [],
  service_choice: [],
  service_remote: [...CONTACT, ...WHEN, ...NOTES],
  service_in_person: [...CONTACT_WITH_PHONE, ...WHEN, ...ADDRESS, ...NOTES],
  event_online: [
    ["attendee_name", "name", true],
    ["ticket_type", "choice", true],
  ],
  event_in_person: [
    ["attendee_name", "name", true],
    ["contact_phone", "phone", false],
    ["ticket_type", "choice", true],
  ],
  booking_remote: [...CONTACT, ...WHEN, ...NOTES],
  booking_in_person: [...CONTACT_WITH_PHONE, ...WHEN, ...ADDRESS, ...NOTES],
};

/**
 * Fields the client answers from the device instead of asking the buyer.
 *
 * Only `timezone`. It is still required, still sent, and still validated
 * server-side — nothing about the contract changes. What changes is who
 * supplies it: `Intl` reports the device zone in the exact IANA form the server
 * accepts, so asking the buyer to type `America/New_York` could only ever make
 * the answer worse. The checkout renders it as context under the chosen time
 * rather than as an input.
 *
 * Kept as a list rather than a boolean on the field because the *reason* is
 * per-field. A key belongs here only when the device's answer is strictly
 * better than a typed one, which is a much narrower claim than "we could guess
 * this" — a shipping address could be guessed too, and must not be.
 */
export const AUTO_FILLED_FIELD_KEYS: readonly string[] = ["timezone"];

export function isAutoFilledField(key: string) {
  return AUTO_FILLED_FIELD_KEYS.includes(key);
}

/** The listing-type pill on the checkout's product summary — what the buyer is
 * buying, which is also why the form below asks what it asks. */
export function fulfillmentTypeLabel(kind: MarketplaceFulfillmentKind): string {
  if (kind === "digital") return "Digital product";
  if (kind.startsWith("event_")) return "Event";
  if (kind.startsWith("booking_")) return "Appointment";
  if (kind.startsWith("service_")) return "Service";
  if (kind === "pickup") return "Local pickup";
  return "Physical item";
}

export function ticketOptions(listing: MarketplaceListing): string[] {
  const metadata = (listing.listing_metadata || {}) as Record<string, unknown>;
  const tickets = metadata.tickets;
  if (!Array.isArray(tickets)) return [];
  return tickets
    .map((ticket) => String((ticket as Record<string, unknown>)?.name || "").trim())
    .filter(Boolean);
}

export function fulfillmentFields(
  kind: MarketplaceFulfillmentKind,
  tickets: readonly string[] = []
): FulfillmentField[] {
  const out: FulfillmentField[] = [];
  for (const [key, type, required] of FIELDS[kind] || []) {
    if (key === "ticket_type") {
      // A seller who published no ticket tiers is selling one undifferentiated
      // admission, so there is nothing to choose between.
      if (!tickets.length) continue;
      out.push({ key, type: "choice", required: true, label: LABELS[key], options: [...tickets] });
      continue;
    }
    out.push({ key, type, required, label: LABELS[key] });
  }
  return out;
}

/** Countries where a street and city alone will not get a parcel delivered. */
const REGION_REQUIRED = ["US", "CA", "AU", "IN", "BR", "MX", "MY", "CN", "AR", "ID"];

/**
 * Countries whose addresses are complete without a postal code. Demanding one
 * universally is how a checkout built in the US rejects a valid address in
 * Ireland or Hong Kong.
 */
const NO_POSTAL_CODE = [
  "AE", "AO", "AG", "AW", "BS", "BZ", "BJ", "BW", "BF", "BI", "CM", "CF", "KM", "CG", "CD", "CK",
  "CI", "DJ", "DM", "GQ", "ER", "FJ", "TF", "GM", "GH", "GD", "GY", "HK", "IE", "JM", "KE", "KI",
  "KP", "LY", "MO", "MW", "ML", "MR", "MU", "MS", "NR", "AN", "NU", "QA", "RW", "KN", "LC", "ST",
  "SC", "SL", "SB", "SO", "SR", "SY", "TZ", "TL", "TK", "TO", "TT", "TV", "UG", "VU", "YE", "ZW",
];

/**
 * The first field the buyer still has to fill, or null when the form is
 * complete. Only enough validation to keep the Continue button honest — the
 * server's answer is the one that decides whether a charge happens.
 */
export function firstMissingFulfillmentField(
  kind: MarketplaceFulfillmentKind,
  tickets: readonly string[],
  values: Record<string, string>
): FulfillmentField | null {
  const fields = fulfillmentFields(kind, tickets);
  for (const field of fields) {
    if (field.required && !String(values[field.key] || "").trim()) return field;
  }
  if (!fulfillmentNeedsAddress(kind)) return null;
  const country = String(values.address_country || "").trim().toUpperCase();
  const byKey = (key: string) => fields.find((field) => field.key === key) || null;
  if (REGION_REQUIRED.includes(country) && !String(values.address_region || "").trim()) {
    return byKey("address_region");
  }
  if (!NO_POSTAL_CODE.includes(country) && !String(values.address_postal_code || "").trim()) {
    return byKey("address_postal_code");
  }
  return null;
}

/**
 * The one kind a whole cart group is asked about. A group settles through a
 * single Stripe session, so it asks a single set of questions — the address if
 * anything in it ships, otherwise the scheduling or pickup answer, otherwise
 * nothing. Mirrors the selection in `marketplace_cart_routes.cart_checkout`.
 */
export function groupFulfillmentKind(
  kinds: readonly MarketplaceFulfillmentKind[]
): MarketplaceFulfillmentKind | "" {
  return (
    kinds.find(fulfillmentNeedsAddress) ||
    kinds.find(isScheduledKind) ||
    kinds.find((kind) => kind === "pickup") ||
    ""
  );
}

/** Whether this kind happens at a time the buyer has to name. */
export function isScheduledKind(kind: MarketplaceFulfillmentKind) {
  return kind.startsWith("service_") || kind.startsWith("booking_") || kind.startsWith("event_");
}

/**
 * Whether buyer and seller end up in the same room.
 *
 * This is the question the escrow/safety presentation is actually about — it is
 * advice about meeting a stranger — and it is NOT the same question as "which
 * progress strip does this order get". `ordersDashboard` used to answer both
 * with one set, which meant an in-person haircut could only get the safety panel
 * by also being described to the buyer as a parcel awaiting collection.
 */
export function isInPersonKind(kind: MarketplaceFulfillmentKind) {
  return (
    kind === "pickup" ||
    kind === "service_in_person" ||
    kind === "booking_in_person" ||
    kind === "event_in_person"
  );
}

/** One-line destination for the review step: where this order is actually going. */
export function fulfillmentDestinationSummary(
  kind: MarketplaceFulfillmentKind,
  values: Record<string, string>
): string {
  if (kind === "digital") return "Delivered to your PulseSoc account";
  if (fulfillmentNeedsAddress(kind)) {
    return [values.address_line1, values.address_line2, values.address_city, values.address_region, values.address_postal_code, values.address_country]
      .map((part) => String(part || "").trim())
      .filter(Boolean)
      .join(", ");
  }
  if (kind === "pickup") return "Collected in person from the seller";
  if (kind === "event_online") return "Joined online";
  if (kind === "event_in_person") return "Attended at the venue";
  const when = [values.scheduled_date, values.scheduled_time, values.timezone]
    .map((part) => String(part || "").trim())
    .filter(Boolean)
    .join(" ");
  return when ? `Online, ${when}` : "Online";
}
